In [1]:
import numpy as np
import matplotlib.pyplot as plt
from keras.models import Sequential
from keras.layers.core import Dense
from keras.utils import np_utils
from keras.datasets import mnist

(XTrain, yTrain), (XTest, yTest) = mnist.load_data()

In [2]:
# Umwalndlung in Vektoren
XTrain = XTrain.reshape(60000, 784)
XTest = XTest.reshape(10000, 784)
# Skalierung durch Dision durch 255
XTrain = XTrain/255
XTest = XTest/255

# Klassenvektoren in binäre Klassenmatrizen
YTrain = np_utils.to_categorical(yTrain, 10)
YTest = np_utils.to_categorical(yTest, 10)

In [3]:
myANN = Sequential()
myANN.add(Dense(80,input_dim=784,activation='relu'))
myANN.add(Dense(40,activation='relu'))
myANN.add(Dense(10,activation='sigmoid'))
myANN.compile(loss='mean_squared_error', optimizer='adam', metrics=['accuracy'])
myANN.fit(XTrain, YTrain, batch_size=24, epochs=10, verbose=True)

Epoch 1/10
2500/2500 [==============================] - 2s 704us/step - loss: 0.0167 - accuracy: 0.9024
Epoch 2/10
2500/2500 [==============================] - 2s 656us/step - loss: 0.0072 - accuracy: 0.9571
Epoch 3/10
2500/2500 [==============================] - 2s 633us/step - loss: 0.0053 - accuracy: 0.9693
Epoch 4/10
2500/2500 [==============================] - 2s 641us/step - loss: 0.0043 - accuracy: 0.9757
Epoch 5/10
2500/2500 [==============================] - 2s 741us/step - loss: 0.0036 - accuracy: 0.9797
Epoch 6/10
2500/2500 [==============================] - 2s 932us/step - loss: 0.0031 - accuracy: 0.9831
Epoch 7/10
2500/2500 [==============================] - 2s 905us/step - loss: 0.0027 - accuracy: 0.9846
Epoch 8/10
2500/2500 [==============================] - 2s 904us/step - loss: 0.0025 - accuracy: 0.9862
Epoch 9/10
2500/2500 [==============================] - 2s 895us/step - loss: 0.0022 - accuracy: 0.9884
Epoch 10/10
2500/2500 [==============================] - 2s 920u

In [4]:
score = myANN.evaluate(XTest, YTest, verbose=False)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test score: 0.0040383106097579
Test accuracy: 0.9768999814987183


In [5]:
import numpy as np
from tensorflow import keras
from tensorflow.keras import layers

(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train / 255
x_test = x_test / 255
x_train = np.expand_dims(x_train, -1)
x_test = np.expand_dims(x_test, -1)


# Klassenvektoren in binäre Klaasenmatrizen
y_train = keras.utils.to_categorical(y_train, 10)
y_test = keras.utils.to_categorical(y_test, 10)

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


In [6]:
model = keras.Sequential(
    [
        keras.Input(shape=(28, 28, 1)),
        layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dropout(0.5),
        layers.Dense(num_classes, activation="softmax"),
    ]
)

model.summary()


Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
flatten (Flatten)            (None, 1600)              0         
_________________________________________________________________
dropout (Dropout)            (None, 1600)              0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)               

In [7]:
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

model.fit(x_train, y_train, batch_size=128, epochs=15, validation_split=0.1)
score = model.evaluate(x_test, y_test, verbose=0)
print("Test loss:", score[0])
print("Test accuracy:", score[1])

Epoch 1/15
422/422 [==============================] - 25s 59ms/step - loss: 0.3593 - accuracy: 0.8900 - val_loss: 0.0763 - val_accuracy: 0.9803
Epoch 2/15
422/422 [==============================] - 26s 62ms/step - loss: 0.1079 - accuracy: 0.9672 - val_loss: 0.0555 - val_accuracy: 0.9845
Epoch 3/15
422/422 [==============================] - 29s 69ms/step - loss: 0.0818 - accuracy: 0.9746 - val_loss: 0.0455 - val_accuracy: 0.9877
Epoch 4/15
422/422 [==============================] - 30s 70ms/step - loss: 0.0696 - accuracy: 0.9786 - val_loss: 0.0423 - val_accuracy: 0.9888
Epoch 5/15
422/422 [==============================] - 28s 65ms/step - loss: 0.0621 - accuracy: 0.9804 - val_loss: 0.0381 - val_accuracy: 0.9895
Epoch 6/15
422/422 [==============================] - 27s 63ms/step - loss: 0.0552 - accuracy: 0.9831 - val_loss: 0.0342 - val_accuracy: 0.9903
Epoch 7/15
422/422 [==============================] - 27s 64ms/step - loss: 0.0513 - accuracy: 0.9839 - val_loss: 0.0337 - val_accuracy:

In [ ]:
import numpy as np
from tensorflow import keras
from keras.datasets import mnist
from sklearn.neighbors import KNeighborsClassifier
(XTrain, yTrain), (XTest, yTest) = mnist.load_data()

# Umwalndlung in Vektoren
XTrain = XTrain.reshape(60000, 784)
XTest = XTest.reshape(10000, 784)
# Skalierung durch Dision durch 255
XTrain = XTrain/255
XTest = XTest/255

knn=KNeighborsClassifier(n_neighbors=20)
knn.fit(XTrain, yTrain)
knn.score(XTest, yTest)

# Resultate
## Fully Connected Netz
Es wurden drei Layer erstellt. Die erste mit 80 Konten, die zweite mit 40, und die dritte mit 10.
Für die ersten bdieden wurde relu Funktion verwendet, für die letzte die sigmoid.
Das fully Connected Netz führt zu einer Genauigkeit von ~97.7% mit einem Loss von 0.004.
## CNN
Für CNN wurden vier Layer verwendet, je zwei Conv2D und und zwei MaxPooling2D. Am Ende wurden noch flatten, dropout und dense Layers verwendet um die Daten zu strukturieren.
Mit dem CNN haben wir eine Genauigkeit von ~99.2% mit einem Loss von 0.026
## KNN
Mit KNN und 10 Nachbarn wurde eine Genauigkeit von 96.65% erreicht. Der Loss ist leider nicht ersichtlich.

## Vergleich
CNN hat die beste Genauigkeit, welche mit 99.2% enorm hoch ist. Das fully connected Netz hat eine etwas tiefere Genauigkeit, aber ein tieferer Loss. Wobei der Loss bei CNN auch nicht hoch ist. Die Genauigkeit von KNN lässt zu wünschen übrig, dies liegt aber auch an der tiefen Anzahl von Nachbarn, welche gewählt wurde.